File: src/registry.rs
// access registry on windows, returns None on non-windows

#[cfg(target_os = "windows")]
extern crate winreg;

#[cfg(target_os = "windows")]
use std::io;

#[cfg(target_os = "windows")]
pub fn hklm(key_name: &str, value_name: &str) -> io::Result<String>
{
    use winreg::enums::*;
    use winreg::RegKey;
    use std::path::Path;

    let hklm = RegKey::predef(HKEY_LOCAL_MACHINE);
    let cur_ver = hklm.open_subkey(key_name)?; // "SOFTWARE\\Microsoft\\Windows\\CurrentVersion"
    let value: String = cur_ver.get_value(value_name)?; // "ProgramFilesDir"
    Ok(value)
}

#[cfg(not(target_os = "windows"))]
pub fn hklm(_: &str, _: & str) -> Result<String, String>
{
    Err(String::from("Not on windows"))
}

File: src/Registry.cs
namespace Workbench;

// access registry on windows, returns None on non-windows

public static class Registry
{
    public static string? hklm(string key_name, string value_name)
    {
        if (Core.IsWindows() == false)
        {
            return null;
        }

        var root = Microsoft.Win32.Registry.LocalMachine;

        var key = root.OpenSubKey(key_name);
        if (key == null) { return null; }

        var kind = key.GetValueKind(value_name);
        if (kind != Microsoft.Win32.RegistryValueKind.String) { return null; }

        var value = key.GetValue(value_name);
        if (value == null) { return null; }

        return value.ToString();
    }
}